pyqtorch/apply.py

```diff
@@ -57,3 +59,32 @@ def apply_operator(
         map(lambda e: "".join(list(e)), [operator_dims, in_state_dims, out_state_dims])
     )
     return einsum(f"{operator_dims},{in_state_dims}->{out_state_dims}", operator, state)
+
+
+def apply_op_op(operator_1: Tensor, operator_2: Tensor, target: int) -> Tensor:
+    """
+    Compute the product of two operators.
+
+    Args:
+        operator_1 (Tensor): The first operator.
+        operator_2 (Tensor): The second operator.
+        target (int): The target qubit index.
+
+    Returns:
+        Tensor: The product of the two operators.
+    """
+
+    n_qubits_1 = int(log2(operator_1.size(1)))
+    n_qubits_2 = int(log2(operator_2.size(1)))
+    batch_size_1 = operator_1.size(-1)
+    batch_size_2 = operator_2.size(-1)
+    if n_qubits_1 > n_qubits_2:
+        operator_2 = promote_op(operator_2, target, n_qubits_1)
+    if n_qubits_1 < n_qubits_2:
+        operator_1 = promote_op(operator_1, target, n_qubits_2)
+    if batch_size_1 > batch_size_2:
+        operator_2 = operator_2.repeat(1, 1, batch_size_1)
```

Comment:
Indeed, there is an issue if we want to switch from a batch size of 2 to 3 because the function will repeat a tensor with `batch_size = 2` three times, thus creating one with `batch_size = 6`. 
However, in our use case of this function, such a scenario is impossible since it involves computing a product between a gate always with `batch_size = 1` and a batch_size-invariant density matrix.

 So, I'm not sure if a change is necessary. What do you think?

Reply:
yeah thats true. i think we can keep it then

Reply:
Finally i fixed it by doing this version:
```py
def operator_product(operator_1: Tensor, operator_2: Tensor, target: int) -> Tensor:
    """
    Compute the product of two operators.

    Args:
        operator_1 (Tensor): The first operator.
        operator_2 (Tensor): The second operator.
        target (int): The target qubit index.

    Returns:
        Tensor: The product of the two operators.
    """

    n_qubits_1 = int(log2(operator_1.size(1)))
    n_qubits_2 = int(log2(operator_2.size(1)))
    batch_size_1 = operator_1.size(-1)
    batch_size_2 = operator_2.size(-1)
    if n_qubits_1 > n_qubits_2:
        operator_2 = promote_operator(operator_2, target, n_qubits_1)
    elif n_qubits_1 < n_qubits_2:
        operator_1 = promote_operator(operator_1, target, n_qubits_2)
    if batch_size_1 > batch_size_2:
        operator_2 = operator_2.repeat(1, 1, batch_size_1)[:,:,:batch_size_1]
    elif batch_size_2 > batch_size_1:
        operator_1 = operator_1.repeat(1, 1, batch_size_2)[:,:,:batch_size_2]


    return batch_last(torch.bmm(batch_first(operator_1), batch_first(operator_2)))
```
But maybe this version is more efficient in term of complexity as this case is rare:
```python3
def operator_product(operator_1: Tensor, operator_2: Tensor, target: int) -> Tensor:
    """
    Compute the product of two operators.

    Args:
        operator_1 (Tensor): The first operator.
        operator_2 (Tensor): The second operator.
        target (int): The target qubit index.

    Returns:
        Tensor: The product of the two operators.
    """

    n_qubits_1 = int(log2(operator_1.size(1)))
    n_qubits_2 = int(log2(operator_2.size(1)))
    batch_size_1 = operator_1.size(-1)
    batch_size_2 = operator_2.size(-1)
    if n_qubits_1 > n_qubits_2:
        operator_2 = promote_operator(operator_2, target, n_qubits_1)
    elif n_qubits_1 < n_qubits_2:
        operator_1 = promote_operator(operator_1, target, n_qubits_2)
    if batch_size_1 > batch_size_2:
        if batch_size_2 == 1:
            operator_2 = operator_2.repeat(1, 1, batch_size_1)
        else:
            operator_2 = operator_2.repeat(1, 1, batch_size_1)[:,:,:batch_size_1]
    elif batch_size_2 > batch_size_1:
        if batch_size_1 == 1:
            operator_1 = operator_1.repeat(1, 1, batch_size_2)
        else:
            operator_1 = operator_1.repeat(1, 1, batch_size_2)[:,:,:batch_size_2]

    return batch_last(torch.bmm(batch_first(operator_1), batch_first(operator_2)))

```
What do you think ?

Reply:
I guess the first one is already quite clear. If it takes into account the second case then it is fine. I would rename `operator_1` -> `op1` since it is quite obvious what we are dealing with.

Reply:
Done.